c/frame/py_frame_names.cc

```diff
@@ -335,4 +328,124 @@ void Frame::_replace_names_from_map(py::odict replacements)
 }
 
 
+/**
+ * Compute Levenshtein distance between two strings `a` and `b`, as described in
+ * https://en.wikipedia.org/wiki/Levenshtein_distance
+ *
+ * Use iterative algorithm, single-row version. The temporary storage required
+ * for the calculations is passed in array `v`, which must be allocated for at
+ * least `min(a.size(), b.size()) + 1` elements.
+ */
+static double _dlevenshtein(
+    const std::string& a, const std::string& b, double* v)
+{
+  const char* aa = a.data();
+  const char* bb = b.data();
+  int n = static_cast<int>(a.size());
+  int m = static_cast<int>(b.size());
+  if (n > m) {
+    std::swap(aa, bb);
+    std::swap(m, n);
+  }
+  // Remove common prefix from the strings
+  while (n && *aa == *bb) {
+    n--; m--;
+    aa++; bb++;
+  }
+  // Remove common suffix from the strings
+  while (n && aa[n - 1] == bb[m - 1]) {
+    n--; m--;
+  }
+  if (n == 0) return m;
+  xassert(0 < n && n <= m);
+  // Compute the Levenshtein distance
+  aa--;  // Shift pointers, so that we can use 1-based indexing below
+  bb--;
+  for (int j = 1; j <= n; ++j) v[j] = j;
+  for (int i = 1; i <= m; ++i) {
+    double w = i - 1;
+    v[0] = i;
+    for (int j = 1; j <= n; ++j) {
+      char ach = aa[j];
+      char bch = bb[i];
+      double c = 0.0;
+      if (ach != bch) {
+        // Use non-trivial cost function to compare character substitution:
+        //   * the cost is lowest when 2 characters differ by case only,
+        //     or if both are "space-like" (i.e. ' ', '_' or '.')
+        //   * medium cost for substituting letters with letters, or digits
+        //     with digits
+        //   * highest cost for all other substitutions.
+        //
```

Comment:
I like the idea to use different costs. Do you know where the values 0.2/0.75/1.0 are coming from?

Reply:
from thin air :)